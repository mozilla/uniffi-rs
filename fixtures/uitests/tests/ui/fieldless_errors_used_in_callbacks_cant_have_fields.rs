// Unfortunately, path is relative to a temporary build directory :-/
uniffi_macros::generate_and_include_scaffolding!("../../../../fixtures/uitests/src/errors.udl");

fn main() { /* empty main required by `trybuild` */}

#[derive(Debug)]
enum ArithmeticError {
  IntegerOverflow,
  // Since this is listed in the UDL as not having fields and is used in a callback interface, it
  // really needs to have no fields.
  DivisionByZero { numerator: u64 },
  // Tuple-style fields are also invalid
  UnexpectedError(String),
}

impl std::fmt::Display for ArithmeticError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        <Self as std::fmt::Debug>::fmt(self, f)
    }
}

impl From<uniffi::UnexpectedUniFFICallbackError> for ArithmeticError {
    fn from(e: uniffi::UnexpectedUniFFICallbackError) -> ArithmeticError {
        ArithmeticError::UnexpectedError(e.to_string())
    }
}

pub trait Calculator {
    fn add(&self, a: u64, b: u64) -> Result<u64, ArithmeticError>;
}
