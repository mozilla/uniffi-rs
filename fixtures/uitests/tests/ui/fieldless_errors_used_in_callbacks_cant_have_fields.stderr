error[E0533]: expected value, found struct variant `Self::DivisionByZero`
 --> $OUT_DIR[uniffi_uitests]/errors.uniffi.rs
  |
  | / #[::uniffi::ffi_converter_error(
  | |     tag = crate::UniFfiTag,
  | |     flat_error,
  | |     with_try_read,
  | | )]
  | |__^ not a value
  |
  = note: this error originates in the attribute macro `::uniffi::ffi_converter_error` (in Nightly builds, run with -Z macro-backtrace for more info)
